### PR TITLE
upgrade formidable to v3

### PR DESCRIPTION
**Pull request by znewton**
# Upgrade Formidable to v3

Taking over #10434, since there are more braking changes than expected.

## Description

The currently used version of ["formidable" package](https://www.npmjs.com/package/formidable/v/3.2.4) has a [vulnerability bug](https://github.com/node-formidable/formidable/issues/856) which has been fixed in the latest version (3.2.4). Bumping up the version to the latest to address the vulnerability issue.

## Steps to Reproduce Bug and Validate Solution

> Only applicable if the work is to address a bug. Please remove this section if the work is for a feature or story
> Provide details on the environment the bug is found, and detailed steps to recreate the bug.
> This should be detailed enough for a team member to confirm that the bug no longer occurs

## PR Checklist

> Use the check-list below to ensure your branch is ready for PR. If the item is not applicable, leave it blank.

-   [ ] I have updated the documentation accordingly.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] My code follows the code style of this project.
-   [x] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

> If this introduces a breaking change, please describe the impact and migration path for existing applications below.

## Testing

> -   Instructions for testing and validation of your code so the reviewer can follow those steps and validate code works as expected

## Any relevant logs or outputs

> -   Use this section to provide either screenshots or output of logs as code snippets

## Other information or known dependencies

> -   Any other information or known dependencies that is important to this PR.
> -   TODO that are to be done after this PR.

### CODE REVIEW

File: server/routerlicious/packages/services-shared/src/restLessServer.ts [hunk 1]
```diff
@@ -45,12 +46,16 @@ export class RestLessServer {
 
     private translateRequestFields(request: IncomingMessageEx, fields: Record<string, any>): void {
         // Parse and override HTTP Method
-        const methodOverride = fields[
+        const methodOverrideField = fields[
             RestLessFieldNames.Method
-        ] as string;
-        request.method = methodOverride;
+        ] as string[] | string;
```
Comment (hedasilv) [inline server/routerlicious/packages/services-shared/src/restLessServer.ts:51]: Curious, was changing the type of variables produced by reading `fields` to `string[] | string` something related to formidable/inclusion?

File: server/routerlicious/packages/services-shared/src/restLessServer.ts [hunk 2]
```diff
@@ -99,8 +108,9 @@ export class RestLessServer {
     }
 
     private async parseStreamRequestFormBody(request: IncomingMessageEx): Promise<void> {
+        const formidable = (await inclusion("formidable")).default;
```
Comment (hedasilv) [inline server/routerlicious/packages/services-shared/src/restLessServer.ts:111]: Does this mean that the `formidable` version we are bumping to does not allow what we used to have, like `import formidable from "formidable";`?

If that is the case, I'm curious: why do we need to use `inclusion` instead of doing something like
```ts
// eslint-disable-next-line @typescript-eslint/no-require-imports
import formidable = require("formidable");
```

File: server/routerlicious/packages/services-shared/src/restLessServer.ts [hunk 3]
```diff
@@ -99,8 +108,9 @@ export class RestLessServer {
...
         return new Promise<void>((resolve, reject) => {
-            const form = formidable({ multiples: true });
+            const form = formidable();
```
Comment (hedasilv) [inline server/routerlicious/packages/services-shared/src/restLessServer.ts:113]: You mentioned in the other closed PR that `multiples` support was removed. How does that affect us?

File: server/routerlicious/packages/services-shared/src/test/restLess.spec.ts [hunk 4]
```diff
@@ -36,17 +35,24 @@ describe("RestLess", () => {
...
-        app.use(restLessMiddleware());
         // set up rudimentary authentication
         const authMiddleware: () => express.RequestHandler = () => (req, res, next) => {
             if (req.get("Authorization") !== `Bearer ${authToken}`) {
                 return res.sendStatus(403);
             }
             next();
         };
-        app.use(authMiddleware());
+        if (restLessBeforeBodyParser) {
```
Comment (hedasilv) [inline server/routerlicious/packages/services-shared/src/test/restLess.spec.ts:45]: Educate me: are we introducing these 2 modes of tests depending on the value of `restLessBeforeBodyParser` just to improve our test coverage, or is this also something related to `formidable`?

**Review by hedasilv — APPROVED** [review]
Thanks for taking care of this! Just left some comments which are really just curious questions to help me understand some details better 😄

File: server/routerlicious/packages/services-shared/src/restLessServer.ts (hunk 1 above)
Comment (znewton) [reply server/routerlicious/packages/services-shared/src/restLessServer.ts:51]: Yes, Formidable now always returns `fields[*]` as an array as of v3.0.0

File: server/routerlicious/packages/services-shared/src/restLessServer.ts (hunk 2 above)
Comment (znewton) [reply server/routerlicious/packages/services-shared/src/restLessServer.ts:111]: This is due to Formidable switching to ESModules in v3.0.0. It's a really annoying problem, where packages that use ESModules cannot be imported using `require()`. You would think that `import formidable as "formidable"` would work, but Typescript is transpiling our code into CommonJS which uses `require` wherever there is an `import` (take a look at the `dist/` folders locally). @tylerbutler helpfully pointed us to a [forum thread with possible fixes](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) (none worked) and the [`inclusion` hack](https://github.com/davidmarkclements/inclusion), which does work.

File: server/routerlicious/packages/services-shared/src/restLessServer.ts (hunk 3 above)
Comment (znewton) [reply server/routerlicious/packages/services-shared/src/restLessServer.ts:113]: Formidable now always returns an array, where it used to be dependent on the number of form fields with that field name (1 field name would return `string`, multiple fieldnames would return `array`). We made use of this expecting multiple Header fields and handling the array possibility. However, for body and method, we expected 1, so a `string`. You can see the type changes above in the code (where you commented already). We still need to handle `string` because if the body is parsed by `bodyparser` instead, it will be a `string | string[]`

File: server/routerlicious/packages/services-shared/src/test/restLess.spec.ts (hunk 4 above)
Comment (znewton) [reply server/routerlicious/packages/services-shared/src/test/restLess.spec.ts:45]: This is to improve test coverage. Only 1 path of RestLess server was being tested (the `formidable` path). So I added this alternate variation to test the bodyparser path. All of our services use the formidable path, which is good, because you can see here I had to make a change to urlencoded to allow it to recognize a restless form body.

File: server/routerlicious/packages/services-shared/src/restLessServer.ts (hunk 2 above)
Comment (hedasilv) [reply server/routerlicious/packages/services-shared/src/restLessServer.ts:111]: Thanks for all the details!
